### DISCUSSION TITLE
Global /renderer/+onBeforeRender and Local /page/+onBeforeRender behavior Vike V1 Client Routing

### DISCUSSION THREAD

**Discussion by JoshShearer** [vikejs/vike#1226] [Discussion]
I have a project where I am using a global +onBeforeRender hook to manage redux state and several local +onBeforeRender hooks for specific pages requiring data fetching.

expected behavior => if page has a local +onBeforeRender hook, run, followed by global +onBeforeRender.  

in my specific case the home page does not contain a +onBeforeRender hook, so the global runs.  When I navigate to  a page with data fetching (local +onBeforeRender) it runs the global hook but misses the local.  

I believe I read in the discussions/documentation that the local should always run?  I would love some clarity!

Thanks.

**Comment by JoshShearer** [comment]
Just realized that I missed the '+' on the onBeforeRender filename so that would explain the global only.  I can now see that global runs, unless local is available, then local only.  Would the appropriate solution here be a custom config?

**Reply by lourot** [reply]
@JoshShearer could you get away with having the local one calling the global one, or would you consider that dirty?

**Reply by brillout** [reply]
@JoshShearer A custom config sounds good to me.

@AurelienLourot My long-term thinking is to position `onBeforeRender()` as the "data orchestrator". In other words: there is only one `onBeforeRender()` per page (which can be defined in user land, but can also be defined by an integration package e.g. `vike-react-query`).

The `onBeforeRender()` hook can then use custom configs. For example a custom `config.query` for GraphQL or `config.endpointURL` for React Query.

FYI @phonzammi.

**Reply by phonzammi** [reply]
I've tried both, they both work. 
1. Having local `onBeforeRender()` per page.  (vike with builtin renderer)
2. One global onBeforeRender and exporting custom `config.query`. (without builtin renderer)

I kinda like the second way.

**Reply by JoshShearer** [reply]
@AurelienLourot This could potentially be a solution, not sure why it would be considered "dirty", but I would like to understand tradeoffs/potential issues with this strategy.  I can just import and run /renderer/+onBeforeRender?

@brillout I attempted a custom config, but hit some snags, is there is more comprehensive V1 example available that I have not found?  I reviewed https://github.com/vikejs/vike/tree/main/examples/custom-preload-v1 but it does not quite seem to replicate what I am trying to do.  

@phonzammi Can you share an example of #2 custom `config.query`?

**Reply by phonzammi** [reply]
> Can you share an example of #2 custom `config.query`?

I haven't finish with the v1 design yet, but the idea is similar to [vps-react-query](https://github.com/phonzammi/vps-react-query)

Hi @JoshShearer, do you notice [this](https://github.com/vikejs/vike-react/discussions/30#discussioncomment-7458066)?

**Reply by lourot** [reply]
> I attempted a custom config, but hit some snags

@JoshShearer would you like elaborate and/or push what you attempted somewhere? Did you notice https://vike.dev/meta ?

**Reply by JoshShearer** [reply]
Hello @AurelienLourot, I have looked through the meta docs.  So far I have worked around using meta, as the examples did not seem to be sophisticated enough to allow me to understand exactly how to use meta for custom hooks/exports.  "title" is simple enough to understand, but I was trying to understand the correct way, as well as how to do something similar to the pre V1 react-query example.    Ultimately Instead of having the export living in Meta, i just put it in +onBeforeRender and have not had problems so far.  

@phonzammi, the discussion was very helpful, thank you.

**Reply by phonzammi** [reply]
@JoshShearer 
> Can you share an example of https://github.com/vikejs/vike/issues/2 custom config.query?

Here [vike-react-tanstack-query](https://github.com/phonzammi/vike-react-tanstack-query)

**Comment by lourot** [comment]
Maybe helpful: we have now invented a new [`data()` hook](https://vike.dev/data), very similar to the `onBeforeRender()` hook. So you can now have a global `onBeforeRender()` hook and a page-specific `data()` hook.
